### PR TITLE
Add text-guided depth-to-image pipeline based on sd-v2-depth model

### PR DESCRIPTION
**Pull request by SamitHuang**
Re-use MiDas depth estimator from @wtomin 
Strength control modified from https://github.com/mindspore-lab/mindone/pull/87

TODO:
- Parallel running on multiple initial images

### CODE REVIEW

File: examples/stable_diffusion_v2/depth_to_image.py [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+        help="path to depth map. If None, depth_map will be extracted from input image",
+    )
+    parser.add_argument("--depth_est_amp_level", type=str, default="O3", help="amp level for running depth estimator")
+    parser.add_argument("--save_path", type=str, default="output/depth", help="path to save image")
+    parser.add_argument("--prompt", type=str, required=True, help="")
+    parser.add_argument("--negative_prompt", type=str, default=None, help="")
+    parser.add_argument("--config", type=str, default=None, help="")
+    parser.add_argument("--ckpt_path", type=str, default=None, help="")
+    parser.add_argument("--aug", type=str, default="resize", help="augment type")
+    parser.add_argument("--mask_ratio", type=float, default=0.75, help="")
```
Comment (wtomin) [inline examples/stable_diffusion_v2/depth_to_image.py:479]: `mask_ratio` is not used in this script.

File: examples/stable_diffusion_v2/scripts/run_train_v2_lora.sh [hunk 2]
```diff
@@ -3,11 +3,10 @@ export HCCL_CONNECT_TIMEOUT=600
 export ASCEND_GLOBAL_LOG_LEVEL=3
 export ASCEND_SLOG_PRINT_TO_STDOUT=0
 
-device_id=1
+device_id=6
```
Comment (wtomin) [inline examples/stable_diffusion_v2/scripts/run_train_v2_lora.sh:6]: device_id should be zero, right?

**Review by wtomin — APPROVED** [review]
Overall it is good to me.

**Review by hadipash — APPROVED** [review]

File: examples/stable_diffusion_v2/conditions/depth/midas.py [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+                    (
+                        "1",
+                        conv_nd(2, neck_dims[3], neck_dims[3], 3, stride=2, padding=1, has_bias=True, pad_mode="pad"),
+                    ),
+                    ("2", conv_nd(2, neck_dims[3], fusion_dim, 3, padding=1, has_bias=False, pad_mode="pad")),
+                ]
+            )
+        )
+        self.fusion4 = FusionBlock(fusion_dim)
+
```
Comment (zhtmike) [inline examples/stable_diffusion_v2/conditions/depth/midas.py:295]: better to clean the commented code

File: examples/stable_diffusion_v2/depth_to_image.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+        rank=0,
+        log_level=eval(args.log_level),
+    )
+
+    # init
+    device_id = int(os.getenv("DEVICE_ID", 0))
+    ms.context.set_context(
+        mode=args.ms_mode,
+        # mode=ms.context.GRAPH_MODE,
+        device_target="Ascend",
```
Comment (zhtmike) [inline examples/stable_diffusion_v2/depth_to_image.py:336]: only for Ascend now?

**Review by zhtmike — APPROVED** [review]

File: examples/stable_diffusion_v2/depth_to_image.py [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+
+    # build models
+    model = load_model_from_config(config, args.ckpt_path)
+
+    # build sampler
+    sname = args.sampler.lower()
+    if sname == "plms":
+        sampler = PLMSSampler(model)
+    # elif sname == 'dpm_solver_pp':
+    #    sampler = DPMSolverSampler(model, "dpmsolver++", prediction_type='noise')
```
Comment (DexterJZ) [inline examples/stable_diffusion_v2/depth_to_image.py:398]: Commented code can be removed

**Review by DexterJZ — APPROVED** [review]

File: examples/stable_diffusion_v2/conditions/depth/midas.py [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+        x = ops.bmm(attn, v.permute(0, 2, 1, 3)).permute(0, 2, 1, 3)
+        x = x.reshape(b, l, c)
+
+        # output
+        x = self.proj(x)
+        return x
+
+
+class GELU(nn.Cell):
+    def __init__(self):
```
[the remark this answers is not in the document]
Comment (wtomin) [reply examples/stable_diffusion_v2/conditions/depth/midas.py:56]: Interestingly I found mindspore.nn.GELU(approximate=False) isn't close enough to `torch.nn.GELU()`. Although Mindspore website claims that [ After testing, the output is more similar to Pytorch when approximate is False](https://www.mindspore.cn/docs/en/r2.1/note/api_mapping/pytorch_diff/GELU.html), I found the self-defined `GELU` class yields better results.

[the remark this answers is not in the document]
Comment (hadipash) [reply examples/stable_diffusion_v2/conditions/depth/midas.py:56]: That's surprising.

File: examples/stable_diffusion_v2/depth_to_image.py [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+    else:
+        # TODO: support midas v3 hybrid
+        raise NotImplementedError
+
+    auto_mixed_precision(depth_model, amp_level=amp_level)
+
+    return depth_model
+
+
+def estimate_depth(images, depth_estimator, amp_level="O2"):
```
[the remark this answers is not in the document]
Comment (SamitHuang) [reply examples/stable_diffusion_v2/depth_to_image.py]: done

**Pull request merged**